compiler/src/main/kotlin/se/ansman/kotshi/DefaultValueProviders.kt

```diff
@@ -0,0 +1,91 @@
+
+class DefaultValueProviders(private val types: Types) {
+    private val providers: MutableList<DefaultValueProvider> = ArrayList()
+
+    fun register(provider: DefaultValueProvider) {
+        providers.add(provider)
+    }
+
+    fun validate() {
+        providers
+                .asSequence()
+                .groupByTo(LinkedHashMap()) { it.type to it.qualifiers }
+                .forEach { (type, qualifiers), grouped ->
+                    // This fix is needed because @JvmStatic causes a static function AND the companion function to be
+                    // generated so we remove the companion function
+                    if (grouped.size > 1) {
+                        val companion = grouped.firstOrNull {
+                            it.element.enclosingElement.simpleName.contentEquals("Companion")
+                        }
+                        if (companion != null) {
+                            grouped.remove(companion)
+                            providers.remove(companion)
+                        }
+                    }
+
+                    if (grouped.size > 1) {
+                        throw ProcessingError("Multiple provides provide values for type $type with qualifiers $qualifiers: ${grouped.map { it.accessor }}", grouped.first().element)
+                    }
+                }
+    }
+
+    operator fun get(property: Property): DefaultValueProvider =
+            get(property, true)
+                    ?: get(property, false)
+                    ?: throw ProcessingError("No default value provider found", property.parameter)
+
+    private fun get(property: Property, onlyExactMatch: Boolean): DefaultValueProvider? {
+        val applicable = providers
+                .asSequence()
+                .filter { property.defaultValueQualifiers == it.qualifiers }
+                .filter {
+                    val rawProviderType = types.erasure(it.typeMirror)
+                    val rawPropertyType = property.rawTypeMirror
+                    // The erased types must match or there is no match
+                    if (onlyExactMatch) {
+                        if (!types.isSameType(rawProviderType, rawPropertyType)) {
+                            return@filter false
+                        }
+                    } else {
+                        if (!types.isAssignable(rawProviderType, rawPropertyType)) {
+                            return@filter false
+                        }
+                    }
+
+                    // If both types are parameterized we need to check the bounds too
+                    if (it.type is ParameterizedTypeName && property.type is ParameterizedTypeName) {
+                        val prov = it.typeMirror as DeclaredType
+                        val prop = property.typeMirror as DeclaredType
+
+                        // Then all type arguments must match
+                        prop.typeArguments.zip(prov.typeArguments) { propTypeArg, provTypeArg ->
```

Comment:
Is it possible that they would have different numbers of type arguments here? Or would that be caught previously? Like say we have `MyStringMap<T> : Map<String, T>` and are comparing `MyStringMap<Int>` with `Map<String, String>` or something.

Reply:
Valid point. I don't think this case will be supported though as the complexity increases a lot